# **Limpeza de dados**

## **Importação dos dados**

* Neste notebook será realizado a limpeza de uma base de dados que traz informações de fichas de registros individuais sobre **_"Síndrome Respiratória Aguda Grave" (SRAG)_** no Brasil no ano de 2025. Com o título de _"2025 - Banco vivo 04/05/2026 - CSV"_, do site **_Gov_**, a base de dados trabalhada aqui contém informações sobre data de internação do paciente, grupo étnico ao qual pertence, município da internação, local da internação, sintomas, etc. 

🧬 Link para o site: https://dados.gov.br/dados/conjuntos-dados/srag-2021-e-2022

In [103]:
import pandas as pd
import numpy as np

df = pd.read_csv('../Dados/INFLUD25-04-05-2026.csv', sep=';', index_col= 'NU_NOTIFIC', engine='python')

* Aparência inicial da base de dados:

In [104]:
df

,DT_NOTIFIC,SEM_NOT,DT_SIN_PRI,SEM_PRI,SG_UF_NOT,ID_REGIONA,CO_REGIONA,ID_MUNICIP,CO_MUN_NOT,CS_SEXO,...,VG_OMS,VG_OMSOUT,VG_LIN,VG_MET,VG_METOUT,VG_DTRES,VG_ENC,VG_REINF,VG_CODEST,REINF
NU_NOTIFIC,,,,,,,,,,,,,,,,,,,,,
31743527606384,2025-03-16T00:00:00.000Z,12,2025-03-13T00:00:00.000Z,11,SP,GVE VII SANTO ANDRE,1332.0,SAO CAETANO DO SUL,354880.0,F,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2
31755001798720,2025-08-09T00:00:00.000Z,32,2025-07-07T00:00:00.000Z,28,RR,NaN,NaN,BOA VISTA,140010.0,F,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2
31743372054067,2025-03-30T00:00:00.000Z,14,2025-03-29T00:00:00.000Z,13,PB,III NRS CAMPINA GRANDE,1421.0,CAMPINA GRANDE,250400.0,F,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2
31740491434244,2025-02-25T00:00:00.000Z,9,2025-02-24T00:00:00.000Z,9,PR,02RS METROPOLITANA,1356.0,CURITIBA,410690.0,F,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2
31750101341771,2025-06-16T00:00:00.000Z,25,2025-06-13T00:00:00.000Z,24,CE,13 CRES TIANGUA,1523.0,VICOSA DO CEARA,231410.0,F,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
31766176626711,2025-12-19T00:00:00.000Z,51,2025-12-04T00:00:00.000Z,49,AP,NaN,NaN,MACAPA,160030.0,F,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2
31765397443940,2025-09-23T00:00:00.000Z,39,2025-09-21T00:00:00.000Z,39,RO,VI GRS PORTO VELHO,1483.0,NOVA MAMORE,110033.0,M,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2
31754329043169,2025-08-04T00:00:00.000Z,32,2025-07-20T00:00:00.000Z,30,GO,PIRINEUS,1792.0,ANAPOLIS,520110.0,M,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2


* Para informações estruturais da tabela é necessário o uso do _info()_.

In [105]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 336188 entries, 31743527606384 to 31744312251987
Columns: 193 entries, DT_NOTIFIC to REINF
dtypes: float64(104), int64(12), object(77)
memory usage: 497.6+ MB


* Através do info() descobrimos que a tabela possui:
    * Mais de 300 mil observações(linhas);
    * 193 variáveis(colunas);
    * Três tipos de dados: 
        - 104 sendo de ponto flutuante; 
        - 12 sendo inteiros;
        - 77 sendo "objects" (o que geralmente trata-se de strings).

* Caso seja uma necessidade ver os nomes das colunas usamos o _info()_ com uma pequena adição:

In [4]:
df.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
Index: 336188 entries, 31743527606384 to 31744312251987
Data columns (total 193 columns):
 #    Column      Dtype  
---   ------      -----  
 0    DT_NOTIFIC  object 
 1    SEM_NOT     int64  
 2    DT_SIN_PRI  object 
 3    SEM_PRI     int64  
 4    SG_UF_NOT   object 
 5    ID_REGIONA  object 
 6    CO_REGIONA  float64
 7    ID_MUNICIP  object 
 8    CO_MUN_NOT  float64
 9    CS_SEXO     object 
 10   DT_NASC     object 
 11   NU_IDADE_N  int64  
 12   TP_IDADE    int64  
 13   COD_IDADE   object 
 14   CS_GESTANT  int64  
 15   CS_RACA     int64  
 16   CS_ETINIA   object 
 17   CS_ESCOL_N  float64
 18   ID_PAIS     object 
 19   CO_PAIS     int64  
 20   SG_UF       object 
 21   ID_RG_RESI  object 
 22   CO_RG_RESI  float64
 23   ID_MN_RESI  object 
 24   CO_MUN_RES  float64
 25   CS_ZONA     float64
 26   NOSOCOMIAL  float64
 27   AVE_SUINO   float64
 28   FEBRE       float64
 29   TOSSE       float64
 30   GARGANTA    float64
 31   DISPNEIA

* Esse primeiro passo serve para apenas se situar em relação a como os dados foram interpretados pelo Jupyter. E as modificações necessárias para a normalização dos dados.

## **Exclusão de dados**

* No notebook não estará explícito o passo a passo para a escolha das colunas que serão excluídas. Foi um processo longo e de análise pessoal, entretanto, o principal critério usado para as exclusões foi: 
    * Colunas pouco interessantes (com muitos dados faltantes ou muito técnicos) para a análise exploratória futura; 

* Inicialmente serão excluídas 72 colunas.

In [106]:
# colunas que serão removidas

df = df.drop(columns=['PAC_COCBO', 'PAC_DSCBO', 'AVE_SUINO', 'OUT_ANIM', 'OUTRO_DES', 'OBES_IMC', 'MORB_DESC', 'FAB_COV_1', 'FAB_COV_2', 'FAB_COVRF', 
                      'FAB_COVRF2', 'FAB_ADIC', 'FAB_RE_BI', 'LOTE_1_COV', 'LOTE_2_COV', 'LOTE_REF', 'LOTE_REF2', 'LOTE_ADIC', 'LOT_RE_BI', 'FNT_IN_COV', 
                      'OUT_ANTIV', 'OUT_TRAT', 'RAIOX_OUT', 'OUT_AMOST', 'FLUBLI_OUT', 'PCR_OUTRO', 'DS_PCR_OUT', 'SOR_OUT', 'RES_IGG', 'RES_IGM', 'RES_IGA', 
                      'DT_RES', 'COD_IDADE', 'TABAG', 'HISTO_VGM', 'PAIS_VGM', 'CO_PS_VGM', 'LO_PS_VGM', 'DT_VGM', 'DT_RT_VGM', 
                      'FAB_ADIC', 'FAB_RE_BI', 'SURTO_SG', 'CO_DETEC', 'VG_OMSOUT', 'VG_LIN', 'VG_MET', 'VG_METOUT', 'VG_DTRES', 'VG_ENC', 
                      'VG_REINF', 'VG_CODEST', 'VG_OMS', 'SEM_NOT', 'CO_RG_RESI', 'CO_MUN_RES', 'CO_PAIS', 'CO_RG_INTE', 'CO_MU_INTE', 
                      'SG_UF', 'DT_RAIOX', 'TOMO_OUT', 'DT_TOMO', 'DT_COLETA', 'DT_RES_AN', 'DS_AN_OUT', 'DT_PCR', 'FLUASU_OUT', 'TP_AM_SOR', 
                      'DT_CO_SOR', 'TP_SOR', 'OUT_SOR', 'SOR_OUT', 'CO_REGIONA', 'CO_MUN_NOT'])

* Pode parecer uma quantidade exacerbada, mas mesmo excluindo 72 colunas ainda restarão 121.

In [107]:
df.columns

Index(['DT_NOTIFIC', 'DT_SIN_PRI', 'SEM_PRI', 'SG_UF_NOT', 'ID_REGIONA',
       'ID_MUNICIP', 'CS_SEXO', 'DT_NASC', 'NU_IDADE_N', 'TP_IDADE',
       ...
       'DOSE_1_COV', 'DOSE_2_COV', 'DOSE_REF', 'DOSE_2REF', 'DOSE_ADIC',
       'DOS_RE_BI', 'TRAT_COV', 'TIPO_TRAT', 'DT_TRT_COV', 'REINF'],
      dtype='object', length=121)

## **Conversão de tipos**

* Ao carregar os dados, o notebook interpretou a maioria deles como _float64_, seguido de _object_ e _int64_, o que não está completamente correto. 
* Observa-se no **_"Dicionário de Dados"_** que não há entradas de valores com pontos flutuantes, também há a questão de haver datas, mas o tipo _datetime_ não ter aparecido no resumo das informações, o que dá a entender que foram interpretados como _object_.

### **float64 → Int64**

* Como mencionado, não há entradas de ponto flutuante na tabela, o que pode ter ocorrido é um problema que o _int64_ do _Numpy_ tem. Ele não aceita valores nulos (NaN), então ao carregar a tabela o _Pandas_ leu um campo NaN em uma coluna _int64_ e a converteu para o tipo _float64_, que tem suporte para NaN.
* O _Pandas_ tem um tipo especial para resolver esse exato problema que é o _Int64_ (**_Integer extension type_**), sua criação é justamente para conseguir interpretar valores nulos.

In [108]:
#conversão de float sem casa decimais para Int64, object para string e manter o NaN como <NA> (versão de nulidade do Pandas)
df = df.convert_dtypes()

* O tipo Float é atualizado para Integer e o tipo Object é atualizado para String.

In [109]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 336188 entries, 31743527606384 to 31744312251987
Columns: 121 entries, DT_NOTIFIC to REINF
dtypes: Int64(85), string(36)
memory usage: 340.2 MB


* Os dados deixam de ter os 0's após a vírgula e se tornam números inteiros.

In [10]:
df.head(3)

,DT_NOTIFIC,DT_SIN_PRI,SEM_PRI,SG_UF_NOT,ID_REGIONA,ID_MUNICIP,CS_SEXO,DT_NASC,NU_IDADE_N,TP_IDADE,...,DOSE_1_COV,DOSE_2_COV,DOSE_REF,DOSE_2REF,DOSE_ADIC,DOS_RE_BI,TRAT_COV,TIPO_TRAT,DT_TRT_COV,REINF
NU_NOTIFIC,,,,,,,,,,,,,,,,,,,,,
31743527606384,2025-03-16T00:00:00.000Z,2025-03-13T00:00:00.000Z,11,SP,GVE VII SANTO ANDRE,SAO CAETANO DO SUL,F,<NA>,57,3,...,22/09/2021,04/06/2021,<NA>,<NA>,<NA>,<NA>,2,<NA>,<NA>,2
31755001798720,2025-08-09T00:00:00.000Z,2025-07-07T00:00:00.000Z,28,RR,<NA>,BOA VISTA,F,1952-01-01T00:00:00.000Z,73,3,...,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>,2
31743372054067,2025-03-30T00:00:00.000Z,2025-03-29T00:00:00.000Z,13,PB,III NRS CAMPINA GRANDE,CAMPINA GRANDE,F,2002-01-01T00:00:00.000Z,23,3,...,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>,2,<NA>,<NA>,2


### **string → datetime**

* Por fim tem a conversão das datas que inicialmente foram lidas como object e agora estão como strings.

In [110]:
#exemplo de uma coluna com datas sendo lida como string
print(df['DT_NASC'].dtype)

string


* A base de dados segue o padrão de começar a maioria dos nomes das colunas de data com 'DT...', então há como filtrar grande parte das colunas e mudar o tipo sem precisar as listar uma a uma.

In [111]:
#filtro que retornará todas colunas que começam com DT
df.filter(regex = r'^DT').head(3)

,DT_NOTIFIC,DT_SIN_PRI,DT_NASC,DT_UT_DOSE,DT_VAC_MAE,DT_DOSEUNI,DT_1_DOSE,DT_2_DOSE,DT_ANTIVIR,DT_INTERNA,DT_ENTUTI,DT_SAIDUTI,DT_EVOLUCA,DT_ENCERRA,DT_DIGITA,DT_TRT_COV
NU_NOTIFIC,,,,,,,,,,,,,,,,
31743527606384,2025-03-16T00:00:00.000Z,2025-03-13T00:00:00.000Z,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>,2025-03-17T00:00:00.000Z,<NA>,<NA>,2025-03-19T00:00:00.000Z,2025-04-17T00:00:00.000Z,2025-04-01T00:00:00.000Z,<NA>
31755001798720,2025-08-09T00:00:00.000Z,2025-07-07T00:00:00.000Z,1952-01-01T00:00:00.000Z,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>,2025-08-07T00:00:00.000Z,<NA>,<NA>,2025-08-11T00:00:00.000Z,2025-08-20T00:00:00.000Z,2025-08-12T00:00:00.000Z,<NA>
31743372054067,2025-03-30T00:00:00.000Z,2025-03-29T00:00:00.000Z,2002-01-01T00:00:00.000Z,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>,2025-03-30T00:00:00.000Z,2025-03-30T00:00:00.000Z,<NA>,2025-03-31T00:00:00.000Z,2025-04-11T00:00:00.000Z,2025-03-30T00:00:00.000Z,<NA>


* As outras colunas de data fazem alusão a datas de quando os pacientes tomaram vacina, ou seja, o restante das colunas começam com 'DOS...'.

In [112]:
df.filter(regex = r'^DOS').head(3)

,DOSE_1_COV,DOSE_2_COV,DOSE_REF,DOSE_2REF,DOSE_ADIC,DOS_RE_BI
NU_NOTIFIC,,,,,,
31743527606384,22/09/2021,04/06/2021,<NA>,<NA>,<NA>,<NA>
31755001798720,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>
31743372054067,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>


* As modificações para as datas são: exclusão dos horários e modificação para datetime.

In [114]:
#seleciona as colunas que começam com 'DT' e 'DOS'.
columns_dt = df.filter(regex=r'^DT').columns
columns_dos = df.filter(regex=r'^DOS').columns

#exclui os horários das colunas 'DT...'.
df[columns_dt] = df[columns_dt].apply(
    lambda col: pd.to_datetime(col, errors='coerce').dt.date
)

#atualiza as colunas que começam com 'DT' para o tipo datetime
df[columns_dt] = df[columns_dt].apply(pd.to_datetime, errors='coerce')

#atualiza as colunas que começam com 'DOS' para o tipo datetime
df[columns_dos] = df[columns_dos].apply(pd.to_datetime, format='%d/%m/%Y', errors='coerce')



In [115]:
df.filter(regex = r'^DT').head(3)

,DT_NOTIFIC,DT_SIN_PRI,DT_NASC,DT_UT_DOSE,DT_VAC_MAE,DT_DOSEUNI,DT_1_DOSE,DT_2_DOSE,DT_ANTIVIR,DT_INTERNA,DT_ENTUTI,DT_SAIDUTI,DT_EVOLUCA,DT_ENCERRA,DT_DIGITA,DT_TRT_COV
NU_NOTIFIC,,,,,,,,,,,,,,,,
31743527606384,2025-03-16,2025-03-13,NaT,NaT,NaT,NaT,NaT,NaT,NaT,2025-03-17,NaT,NaT,2025-03-19,2025-04-17,2025-04-01,NaT
31755001798720,2025-08-09,2025-07-07,1952-01-01,NaT,NaT,NaT,NaT,NaT,NaT,2025-08-07,NaT,NaT,2025-08-11,2025-08-20,2025-08-12,NaT
31743372054067,2025-03-30,2025-03-29,2002-01-01,NaT,NaT,NaT,NaT,NaT,NaT,2025-03-30,2025-03-30,NaT,2025-03-31,2025-04-11,2025-03-30,NaT


In [116]:
df.filter(regex = r'^DOS').head(3)

,DOSE_1_COV,DOSE_2_COV,DOSE_REF,DOSE_2REF,DOSE_ADIC,DOS_RE_BI
NU_NOTIFIC,,,,,,
31743527606384,2021-09-22,2021-04-06,NaT,NaT,NaT,NaT
31755001798720,NaT,NaT,NaT,NaT,NaT,NaT
31743372054067,NaT,NaT,NaT,NaT,NaT,NaT


## **Modificações adicionais**

In [117]:
df.describe()

,DT_NOTIFIC,DT_SIN_PRI,SEM_PRI,DT_NASC,NU_IDADE_N,TP_IDADE,CS_GESTANT,CS_RACA,CS_ESCOL_N,CS_ZONA,...,DOSE_1_COV,DOSE_2_COV,DOSE_REF,DOSE_2REF,DOSE_ADIC,DOS_RE_BI,TRAT_COV,TIPO_TRAT,DT_TRT_COV,REINF
count,336188,336188,336188.0,336116,336188.0,336188.0,336188.0,336188.0,217310.0,313744.0,...,151429,52603,73961,17915,6192,15272,269096.0,909.0,828,336188.0
mean,2025-07-04 20:29:58.051090688,2025-06-28 02:48:34.198008320,25.955219,1996-05-11 02:39:29.096383616,30.013424,2.718601,5.82726,3.040082,4.685854,1.16604,...,2021-09-14 13:23:11.012289536,2021-10-04 03:01:14.938691840,2022-01-21 20:19:42.006733056,2022-07-09 20:29:24.331565312,2022-03-16 07:29:18.139534592,2023-07-15 21:23:45.667889152,2.570964,2.520352,2025-07-03 01:32:10.434782464,2.0
min,2024-12-29 00:00:00,2024-12-29 00:00:00,1.0,1883-05-06 00:00:00,-1.0,1.0,0.0,1.0,0.0,1.0,...,2018-04-27 00:00:00,2018-08-06 00:00:00,2021-01-16 00:00:00,2021-06-12 00:00:00,2021-01-20 00:00:00,2022-01-04 00:00:00,1.0,1.0,2024-12-30 00:00:00,2.0
25%,2025-05-06 00:00:00,2025-04-29 00:00:00,18.0,1961-04-27 00:00:00,3.0,2.0,6.0,1.0,1.0,1.0,...,2021-03-23 00:00:00,2021-04-08 00:00:00,2021-11-04 00:00:00,2022-04-07 00:00:00,2021-11-22 00:00:00,2023-03-05 00:00:00,2.0,2.0,2025-04-02 18:00:00,2.0
50%,2025-06-22 00:00:00,2025-06-15 00:00:00,24.0,2018-07-14 00:00:00,9.0,3.0,6.0,4.0,5.0,1.0,...,2021-05-11 00:00:00,2021-08-04 00:00:00,2021-12-23 00:00:00,2022-07-04 00:00:00,2022-02-16 00:00:00,2023-07-03 00:00:00,2.0,3.0,2025-07-10 00:00:00,2.0
75%,2025-09-11 00:00:00,2025-09-04 00:00:00,36.0,2024-07-31 00:00:00,64.0,3.0,6.0,4.0,9.0,1.0,...,2021-09-09 00:00:00,2021-11-08 00:00:00,2022-03-21 00:00:00,2022-10-05 00:00:00,2022-05-10 06:00:00,2023-10-05 00:00:00,2.0,3.0,2025-09-23 00:00:00,2.0
max,2026-05-03 00:00:00,2026-01-03 00:00:00,52.0,2026-01-01 00:00:00,141.0,3.0,9.0,9.0,9.0,9.0,...,2026-01-16 00:00:00,2026-07-01 00:00:00,2025-12-09 00:00:00,2025-11-04 00:00:00,2025-10-18 00:00:00,2025-12-06 00:00:00,9.0,3.0,2026-01-22 00:00:00,2.0
std,NaN,NaN,12.763225,NaN,32.433427,0.481504,0.635717,2.202966,3.389927,0.815712,...,NaN,NaN,NaN,NaN,NaN,NaN,1.931439,0.833543,NaN,0.0


* A tabela do _describe_ mostra um erro na coluna 'NU_IDADE_N' que registra a idade dos pacientes, o menor valor é -1. Para garantir que esse é o único valor incorreto nessa coluna basta uma linha de código para conferir:

In [118]:
#retorna todas as ocorrências onde os valores das idades são menores que 0
df[df['NU_IDADE_N'] < 0]['NU_IDADE_N']

NU_NOTIFIC
31746452307100    -1
Name: NU_IDADE_N, dtype: Int64

* É fácil corrigir essa observação, pois basta avaliar as variáveis DT_NASC (data de nascimento), DT_SIN_PRI (data dos primeiros sintomas) e TP_IDADE (tipo idade).

In [119]:
df.loc[31746452307100, ['DT_SIN_PRI', 'DT_NASC', 'TP_IDADE', 'NU_IDADE_N']]

DT_SIN_PRI    2025-04-24 00:00:00
DT_NASC       2025-04-10 00:00:00
TP_IDADE                        2
NU_IDADE_N                     -1
Name: 31746452307100, dtype: object

* Nessa observação vemos que a idade por algum motivo foi classificada como do Tipo 2 (mês) sendo que o correto, de acordo com o dicionário de dados, seria ser classificada como do Tipo 1 (dias), como é algo simples de ser corrigido não optarei pela exclusão da linha, apenas farei algumas modificações nas colunas NU_IDADE_N e TP_IDADE.
* No dicionário o cálculo da idade é baseado em _data dos primeiros sintomas_ - _data do nascimento_, e por esse cálculo, nessa observação específica, fazer alusão a um recém-nascido, com menos de um mês de vida, o tipo de idade dele deve ser classificado como 1(dias).

In [120]:
#mudança da idade para 14 (dias) e o tipo para 1.
df.loc[31746452307100, ['NU_IDADE_N', 'TP_IDADE']] = [14, 1]

In [121]:
df.loc[31746452307100, ['DT_SIN_PRI', 'DT_NASC', 'TP_IDADE', 'NU_IDADE_N']]

DT_SIN_PRI    2025-04-24 00:00:00
DT_NASC       2025-04-10 00:00:00
TP_IDADE                        1
NU_IDADE_N                     14
Name: 31746452307100, dtype: object

* Por fim, apenas para garantir, conferirei inconsistências nas idades registradas com base no tipo nos quais elas foram classificadas.

In [123]:
#filtros para separar as idades de acordo com o tipo (dias, meses, anos)
filter_day = df['TP_IDADE'] == 1
filter_month = df['TP_IDADE'] == 2
filter_year = df['TP_IDADE'] == 3

#seleção das linhas que pertencem a cada categoria 
df_days = df.loc[filter_day].copy()
df_months = df.loc[filter_month].copy()
df_years = df.loc[filter_year].copy()

# subtração da data dos primeiros sintomas menos a data que contém o nascimento em cada observação 
df_days['IDADE_CALCULADA_DIA'] = (
    (df_days['DT_SIN_PRI'] - df_days['DT_NASC']).dt.days
)

df_months['IDADE_CALCULADA_MES'] = (
    (df_months['DT_SIN_PRI'] - df_months['DT_NASC']).dt.days // 30
)

df_years['IDADE_CALCULADA_ANO'] = (
    (df_years['DT_SIN_PRI'] - df_years['DT_NASC']).dt.days // 365
)

# criação das variáveis para visualizar comparações 

difference_days = df_days.loc[
    df_days['IDADE_CALCULADA_DIA'] != df_days['NU_IDADE_N'],
    ['DT_NASC', 'DT_SIN_PRI', 'NU_IDADE_N', 'IDADE_CALCULADA_DIA']
]

difference_months = df_months.loc[
    df_months['IDADE_CALCULADA_MES'] != df_months['NU_IDADE_N'],
    ['DT_NASC', 'DT_SIN_PRI', 'NU_IDADE_N', 'IDADE_CALCULADA_MES']
]

difference_years = df_years.loc[
    df_years['IDADE_CALCULADA_ANO'] != df_years['NU_IDADE_N'],
    ['DT_NASC', 'DT_SIN_PRI', 'NU_IDADE_N', 'IDADE_CALCULADA_ANO']
]


In [29]:
difference_days

,DT_NASC,DT_SIN_PRI,NU_IDADE_N,IDADE_CALCULADA_DIA
NU_NOTIFIC,,,,
31748092424499,2025-05-20,2025-05-21,0,1.0
31748521764196,2025-03-25,2025-05-25,0,61.0
31746617574805,2025-04-13,2025-05-01,19,18.0
31752595735889,2025-06-26,2025-07-09,0,13.0
31764336526404,2025-11-03,2025-11-04,0,1.0
31753101092199,2025-07-07,2025-07-15,14,8.0
31753104664400,2025-07-04,2025-07-16,14,12.0
31742844074686,2025-03-08,2025-03-10,16,2.0
31745936284768,2025-04-12,2025-04-26,9,14.0


In [30]:
difference_months

,DT_NASC,DT_SIN_PRI,NU_IDADE_N,IDADE_CALCULADA_MES
NU_NOTIFIC,,,,
31751419726774,2025-05-16,2025-06-03,1,0.0
31740137473151,2025-01-22,2025-02-17,1,0.0
31750946647431,2025-05-22,2025-06-04,1,0.0
31747245024752,2024-11-28,2025-05-08,6,5.0
31743019926096,2024-09-30,2025-03-21,6,5.0
...,...,...,...,...
31766776071579,2025-11-22,2025-12-20,1,0.0
31766776494593,2025-11-24,2025-12-21,1,0.0
31773414936570,2025-07-27,2025-12-14,5,4.0


In [31]:
difference_years

,DT_NASC,DT_SIN_PRI,NU_IDADE_N,IDADE_CALCULADA_ANO
NU_NOTIFIC,,,,
31761054946537,1962-10-18,2025-10-15,62,63.0
31748959536662,1941-06-15,2025-05-25,83,84.0
31751651298236,1944-07-11,2025-06-23,80,81.0
31744294481213,2019-04-30,2025-04-06,6,5.0
31750948583207,1943-06-24,2025-06-14,81,82.0
...,...,...,...,...
31762436513437,1955-11-06,2025-10-27,69,70.0
31736862359944,1932-01-10,2024-12-29,92,93.0
31774273790420,1940-12-24,2025-12-19,84,85.0


* Apesar de haver discrepâncias nos tipos 2 e 3 (mês, ano), elas são de apenas um número, ou seja, não são tão "fora da curva". Entretanto, quando se trata do tipo 1 (dia), os valores tendem a ser muito diferentes entre si, então vou focar apenas nesse último e fazer as modificações necessárias. Há três tipos de atualizações que precisam ser feitas:
    * 1° - idades que pertencem ao tipo 1, mas por algum motivo o sistema calculou erroneamente seus valores;
    * 2° - idades que partencem ao tipo 2, porém estão registrados como do tipo 1 (campos que possuem valores acima de 29 dias e abaixo de 364);
    * 3° - idades que pertencem ao tipo 3, porém estão registrados como do tipo 1 (campos que possuem valores acima de 364 dias).

In [ ]:
# correção das idades que pertencem ao tipo 1, mas que apresentam erros no registro
real_age = (
    (df['DT_SIN_PRI'] - df['DT_NASC']).dt.days
)

filtro = (
    (df['TP_IDADE'] == 1) &
    (real_age < 30) &
    (df['NU_IDADE_N'] != real_age)
)

df.loc[filtro, 'NU_IDADE_N'] = (
    real_age[filtro]
).astype(int)

In [126]:
df.loc[31738260806730, ['NU_IDADE_N']]

NU_IDADE_N    18
Name: 31738260806730, dtype: object

In [127]:
# correção das idade que estão registradas como tipo 1, mas pertencem ao tipo 2
filtro = (
    (df['TP_IDADE'] == 1) &
    (
        ((df['DT_SIN_PRI'] - df['DT_NASC']).dt.days > 29) &
        ((df['DT_SIN_PRI'] - df['DT_NASC']).dt.days < 365)
    )
)

# Atualizar TP_IDADE
df.loc[filtro, 'TP_IDADE'] = 2

# Atualizar NU_IDADE_N
df.loc[filtro, 'NU_IDADE_N'] = (
    np.ceil(
        ((df.loc[filtro, 'DT_SIN_PRI'] -
          df.loc[filtro, 'DT_NASC']).dt.days) / 30
    )
).astype(int)

In [128]:
df.loc[31753292927733, ['TP_IDADE', 'NU_IDADE_N']]

TP_IDADE       2
NU_IDADE_N    13
Name: 31753292927733, dtype: object

In [129]:
# correção das idade que estão registradas como tipo 1, mas pertencem ao tipo 3
filtro = (
    (df['TP_IDADE'] == 1) &
    (
        (df['DT_SIN_PRI'] - df['DT_NASC']).dt.days > 364
    )
)

# Atualizar TP_IDADE
df.loc[filtro, 'TP_IDADE'] = 3

# Atualizar NU_IDADE_N
df.loc[filtro, 'NU_IDADE_N'] = (
    np.ceil(
        ((df.loc[filtro, 'DT_SIN_PRI'] -
          df.loc[filtro, 'DT_NASC']).dt.days) // 365
    )
).astype(int)

In [130]:
df.loc[31762945347436, ['TP_IDADE', 'NU_IDADE_N']]

TP_IDADE      3
NU_IDADE_N    1
Name: 31762945347436, dtype: object

In [131]:
df.describe()

,DT_NOTIFIC,DT_SIN_PRI,SEM_PRI,DT_NASC,NU_IDADE_N,TP_IDADE,CS_GESTANT,CS_RACA,CS_ESCOL_N,CS_ZONA,...,DOSE_1_COV,DOSE_2_COV,DOSE_REF,DOSE_2REF,DOSE_ADIC,DOS_RE_BI,TRAT_COV,TIPO_TRAT,DT_TRT_COV,REINF
count,336188,336188,336188.0,336116,336188.0,336188.0,336188.0,336188.0,217310.0,313744.0,...,151429,52603,73961,17915,6192,15272,269096.0,909.0,828,336188.0
mean,2025-07-04 20:29:58.051090688,2025-06-28 02:48:34.198008320,25.955219,1996-05-11 02:39:29.096383616,30.011779,2.718687,5.82726,3.040082,4.685854,1.16604,...,2021-09-14 13:23:11.012289536,2021-10-04 03:01:14.938691840,2022-01-21 20:19:42.006733056,2022-07-09 20:29:24.331565312,2022-03-16 07:29:18.139534592,2023-07-15 21:23:45.667889152,2.570964,2.520352,2025-07-03 01:32:10.434782464,2.0
min,2024-12-29 00:00:00,2024-12-29 00:00:00,1.0,1883-05-06 00:00:00,0.0,1.0,0.0,1.0,0.0,1.0,...,2018-04-27 00:00:00,2018-08-06 00:00:00,2021-01-16 00:00:00,2021-06-12 00:00:00,2021-01-20 00:00:00,2022-01-04 00:00:00,1.0,1.0,2024-12-30 00:00:00,2.0
25%,2025-05-06 00:00:00,2025-04-29 00:00:00,18.0,1961-04-27 00:00:00,3.0,2.0,6.0,1.0,1.0,1.0,...,2021-03-23 00:00:00,2021-04-08 00:00:00,2021-11-04 00:00:00,2022-04-07 00:00:00,2021-11-22 00:00:00,2023-03-05 00:00:00,2.0,2.0,2025-04-02 18:00:00,2.0
50%,2025-06-22 00:00:00,2025-06-15 00:00:00,24.0,2018-07-14 00:00:00,9.0,3.0,6.0,4.0,5.0,1.0,...,2021-05-11 00:00:00,2021-08-04 00:00:00,2021-12-23 00:00:00,2022-07-04 00:00:00,2022-02-16 00:00:00,2023-07-03 00:00:00,2.0,3.0,2025-07-10 00:00:00,2.0
75%,2025-09-11 00:00:00,2025-09-04 00:00:00,36.0,2024-07-31 00:00:00,64.0,3.0,6.0,4.0,9.0,1.0,...,2021-09-09 00:00:00,2021-11-08 00:00:00,2022-03-21 00:00:00,2022-10-05 00:00:00,2022-05-10 06:00:00,2023-10-05 00:00:00,2.0,3.0,2025-09-23 00:00:00,2.0
max,2026-05-03 00:00:00,2026-01-03 00:00:00,52.0,2026-01-01 00:00:00,141.0,3.0,9.0,9.0,9.0,9.0,...,2026-01-16 00:00:00,2026-07-01 00:00:00,2025-12-09 00:00:00,2025-11-04 00:00:00,2025-10-18 00:00:00,2025-12-06 00:00:00,9.0,3.0,2026-01-22 00:00:00,2.0
std,NaN,NaN,12.763225,NaN,32.434157,0.481298,0.635717,2.202966,3.389927,0.815712,...,NaN,NaN,NaN,NaN,NaN,NaN,1.931439,0.833543,NaN,0.0


* Diante desses valores negativos inesperados na coluna 'NU_IDADE_N' não faz mal conferir todas as outras colunas.

In [132]:
# verifica todas as colunas numéricas
negative = (df.select_dtypes(include=['Int64']) < 0).any()

# retorna apenas as colunas que têm pelo menos um valor negativo
negative_columns = negative[negative].index.tolist()
print(negative_columns)

[]


* Como há a negativa de valores menores que 0 nas outras colunas numéricas seguiremos para o próximo passo.

## **Criação das tabelas atualizadas**

* De primeiro momento a tabela que foi limpa nesse notebook será salva em outra pasta para a análise exploratória, entretanto, haverá uma subdivisão dela com informações apenas de indígenas. O que gerará, portanto, duas tabelas distintas.

In [133]:
#filtro para trazer apenas as observações que referem-se aos indígenas
indigenas = df[df["CS_RACA"] == 5].copy()

In [134]:
indigenas

,DT_NOTIFIC,DT_SIN_PRI,SEM_PRI,SG_UF_NOT,ID_REGIONA,ID_MUNICIP,CS_SEXO,DT_NASC,NU_IDADE_N,TP_IDADE,...,DOSE_1_COV,DOSE_2_COV,DOSE_REF,DOSE_2REF,DOSE_ADIC,DOS_RE_BI,TRAT_COV,TIPO_TRAT,DT_TRT_COV,REINF
NU_NOTIFIC,,,,,,,,,,,,,,,,,,,,,
31755001798720,2025-08-09,2025-07-07,28,RR,<NA>,BOA VISTA,F,1952-01-01,73,3,...,NaT,NaT,NaT,NaT,NaT,NaT,<NA>,<NA>,NaT,2
31747245024752,2025-05-13,2025-05-08,19,AP,<NA>,MACAPA,F,2024-11-28,6,2,...,NaT,NaT,NaT,NaT,NaT,NaT,2,<NA>,NaT,2
31770814981494,2025-12-14,2025-12-14,51,SP,GVE XXIII REGISTRO,REGISTRO,M,2025-07-24,5,2,...,NaT,NaT,NaT,NaT,NaT,NaT,<NA>,<NA>,NaT,2
31737500007961,2025-01-19,2025-01-15,3,RR,<NA>,BOA VISTA,M,2024-12-13,1,2,...,NaT,NaT,NaT,NaT,NaT,NaT,2,<NA>,NaT,2
31738959856748,2025-02-07,2025-02-04,6,AM,ALTO SOLIMOES,TABATINGA,F,2024-12-14,2,2,...,NaT,NaT,NaT,NaT,NaT,NaT,2,<NA>,NaT,2
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
31747139224600,2025-05-12,2025-05-09,19,DF,<NA>,BRASILIA,F,2000-01-22,25,3,...,2021-08-17,2021-11-11,2022-03-06,NaT,NaT,NaT,2,<NA>,NaT,2
31753883957362,2025-07-20,2025-07-18,29,MS,NRS DE AQUIDAUANA,AQUIDAUANA,F,1937-01-24,88,3,...,2021-01-24,NaT,NaT,NaT,NaT,NaT,2,<NA>,NaT,2
31753882875232,2025-07-13,2025-07-09,28,MS,NRS DE AQUIDAUANA,AQUIDAUANA,F,1969-02-25,56,3,...,NaT,NaT,NaT,NaT,NaT,NaT,2,<NA>,NaT,2


In [136]:
#tabela geral
df.to_parquet('../dados/df_SRAG2025.parquet', index = False)

#tabela sobre indigenas 
indigenas.to_parquet('../dados/df_SRAG2025_indigenas.parquet', index = False)